Building the SmartTag models
===

# Initial checks

In [6]:
source .venv/bin/activate

(.venv) 

: 1

In [1]:
neo4j start

Starting Neo4j.
Started neo4j (pid 59966). It is available at http://localhost:7474/
There may be a short delay until the server is ready.
See /Users/lemberger/Documents/code/neo4j/logs/neo4j.log for current status.


In [2]:
neo4j status

Neo4j is running at pid 59966


In [4]:
ls data

1X_L1200_disease_w_test.zip
1X_L1200_disease_w_train.zip
1X_L1200_ncbi_disease_w_test.zip
1X_L1200_ncbi_disease_w_train.prov
1X_L1200_ncbi_disease_w_train.pyth
1X_L1200_ncbi_disease_w_train.txt
1X_L1200_ncbi_disease_w_train.zip
1X_L1200_ncbi_disease_w_train_textcoded.pyth
5X_L1200_ncbi_disease_w_test.prov
5X_L1200_ncbi_disease_w_test.pyth
5X_L1200_ncbi_disease_w_test.txt
5X_L1200_ncbi_disease_w_test.zip
5X_L1200_ncbi_disease_w_test_textcoded.pyth
5X_L1200_ncbi_disease_w_train.prov
5X_L1200_ncbi_disease_w_train.pyth
5X_L1200_ncbi_disease_w_train.txt
5X_L1200_ncbi_disease_w_train.zip
5X_L1200_ncbi_disease_w_train_textcoded.pyth
5X_L1200_small_molecule_w_test.zip
5X_L1200_small_molecule_w_train.zip
CALBC_SSC_DISO
NCBI_disease
test_brat
test_entities_test.zip
test_entities_train.prov
test_entities_train.pyth
test_entities_train.txt
test_entities_train.zip
test_entities_train_textcoded.pyth
testing_brat_test.zip
testing_brat_train.prov
testing_brat_train.pyth
testing_brat_train.txt
testing_

In [11]:
ls models

ls: models: No such file or directory


: 1

In [ ]:
ls runs

In [8]:
python -m smtag.datagen.sdgraph2th --help

usage: sdgraph2th.py [-h] [-f FILENAMEBASE] [-X ITERATIONS] [-v] [-L LENGTH]
                     [-t TESTFRACT] [-W] [-S] [-d] [-p PADDING]
                     [-A TAGS2ANONYMIZE] [-AA DONOTANONYMIZE] [-l LIMIT]
                     [-Y YEAR_RANGE] [-J JOURNALS] [-D DOI] [-y TYPE] [-e]
                     [-s] [-r ROLE] [-N] [-w WORKING_DIRECTORY]

Generates text and tensor files from tagged text in sd-graph.

optional arguments:
  -h, --help            show this help message and exit
  -f FILENAMEBASE, --filenamebase FILENAMEBASE
                        namebase to save trainset and features set files
  -X ITERATIONS, --iterations ITERATIONS
                        number of times each example is sampled
  -v, --verbose         verbosity
  -L LENGTH, --length LENGTH
                        length of the text snippets used as example
  -t TESTFRACT, --testfract TESTFRACT
                        fraction of papers in testset
  -W, --window          switches to the sampling fig legend

: 1

In [9]:
python -m smtag.train.meta --help

usage: meta.py [-h] [-f FILE] [-E EPOCHS] [-Z MINIBATCH_SIZE]
               [-R LEARNING_RATE] [-V VALIDATION_FRACTION]
               [-o OUTPUT_FEATURES] [-i FEATURES_AS_INPUT]
               [-a OVERLAP_FEATURES] [-c COLLAPSED_FEATURES] [-n NF_TABLE]
               [-k KERNEL_TABLE] [-p POOL_TABLE] [-w WORKING_DIRECTORY]

Top level module to manage training.

optional arguments:
  -h, --help            show this help message and exit
  -f FILE, --file FILE  Namebase of dataset to import (default:
                        test_entities_train)
  -E EPOCHS, --epochs EPOCHS
                        Number of training epochs. (default: 120)
  -Z MINIBATCH_SIZE, --minibatch_size MINIBATCH_SIZE
                        Minibatch size. (default: 128)
  -R LEARNING_RATE, --learning_rate LEARNING_RATE
                        Learning rate. (default: 0.001)
  -V VALIDATION_FRACTION, --validation_fraction VALIDATION_FRACTION
                        Fraction of the dataset that should be used as
 

: 1

In [5]:
python -m smtag.train.meta -Z2 -E1

opt[namebase]=test_entities_train
opt[learning_rate]=0.001
opt[epochs]=1
opt[minibatch_size]=2
opt[selected_features]=['geneprod']
opt[collapsed_features]=[]
opt[overlap_features]=[]
opt[features_as_input]=[]
opt[nf_table]=[8, 8, 8]
opt[pool_table]=[2, 2, 2]
opt[kernel_table]=[6, 6, 6]
opt[dropout]=0.1
opt[validation_fraction]=0.2
nf.output= 1
nf.input= 32
index_of_collapsed_feature= None
index_of_overlap_feature None
concept2index self.selected_features [21]
concept2index self.collapsed_features []
concept2index self.overlap_features []
Loading ./data/test_entities_train.pyth as features for the dataset.
Loading ./data/test_entities_train_textcoded.pyth as encoded text for the dataset.
Loading ./data/test_entities_train.txt for the original texts of the dataset.
Loading ./data/test_entities_train.prov as provenance info for the examples in the dataset.
Dataset dimensions:
100 text examples of size 160
0 input features (in-channels).
22 output features (out-channels).
Creating dataset 

In [1]:
tensorboard --logdir runs &

[1] 59789


In [4]:
ps -a | grep tensoboard

59821 ttys002    0:00.00 grep tensoboard


# A. Data generation

## Small molecules

Generate the data with `-5X` sampling of each example of length `-L1200` characters with a random window method.

In [3]:
python -m smtag.datagen.sdgraph2th -l1000 -L1200 -X5 -y molecule -f 5X_L1200_small_molecule

limit_clause  LIMIT 20 
entity_type_clause AND ( t.type = 'molecule') 
tags2anonmymize_clause 
donotanonymize_clause 
b'0 panels found for figure Figure 1 (1) in paper 10.1289/ehp.84589'
b'2 panels found for figure Figure 2 (53) in paper 10.1289/ehp.84589'
b'4 panels found for figure Figure 3 (89) in paper 10.1289/ehp.84589'
b'0 panels found for figure Figure 4 (127) in paper 10.1289/ehp.84589'
b'3 panels found for figure Figure 5 (154) in paper 10.1289/ehp.84589'
b'0 panels found for figure Figure 6 (178) in paper 10.1289/ehp.84589'
number of figures in  0 10.1289/ehp.84589 3
counted 9 panels.
b'0 panels found for figure Figure 1 (194) in paper 10.1002/eji.200323730'
b'1 panels found for figure Figure 2 (238) in paper 10.1002/eji.200323730'
b'2 panels found for figure Figure 3 (272) in paper 10.1002/eji.200323730'
b'2 panels found for figure Figure 4 (300) in paper 10.1002/eji.200323730'
b'4 panels found for figure Figure 5 (352) in paper 10.1002/eji.200323730'
b'2 panels found for fi

number of figures in  3837 10.1002/embr.201438501 4
counted 152 panels.
b'0 panels found for figure Figure 1 (4294) in paper 10.1002/emmm.201201443'
b'0 panels found for figure Figure 2 (4317) in paper 10.1002/emmm.201201443'
b'0 panels found for figure Figure 3 (4348) in paper 10.1002/emmm.201201443'
b'0 panels found for figure Figure 4 (4367) in paper 10.1002/emmm.201201443'
b'0 panels found for figure Figure 5 (4400) in paper 10.1002/emmm.201201443'
b'0 panels found for figure Figure 6 (4429) in paper 10.1002/emmm.201201443'
b'0 panels found for figure Figure 7 (4446) in paper 10.1002/emmm.201201443'
number of figures in  4293 10.1002/emmm.201201443 0
counted 152 panels.
b'6 panels found for figure Figure 1 (4478) in paper 10.1002/emmm.201303356'
b'1 panels found for figure Figure 2 (4538) in paper 10.1002/emmm.201303356'
b'2 panels found for figure Figure 3 (4573) in paper 10.1002/emmm.201303356'
b'1 panels found for figure Figure 4 (4607) in paper 10.1002/emmm.201303356'
b'0 panel

## Gene and proteins

In [ ]:
python -m smtag.datagen.sdgraph2th -l1000 -L1200 -X5 -y protein,gene -f 5X_L1200_protein_gene

## Cellular components

In [ ]:
python -m smtag.datagen.sdgraph2th -l1000 -L1200 -X10 -y subcellular -f 10X_L1200_subcell

## Cells

In [ ]:
python -m smtag.datagen.sdgraph2th -l1000 -L1200 -X5 -y cell -f 5X_L1200_cell

## Tissue & Organs

In [ ]:
python -m smtag.datagen.sdgraph2th -l1000 -L1200 -X5 -y tissue -f 5X_L1200_tissue

## Organisms

In [ ]:
python -m smtag.datagen.sdgraph2th -l1000 -L1200 -X5 -y organism -f 5X_L1200_organism

## Experimental assays

In [ ]:
python -m smtag.datagen.sdgraph2th -l1000 -L1200 -X5 -y exp_assay -f 5X_L1200_exp_assay

## Diseases
Using a corpus in the brat format. Note that we need to use the module `ann2th`.

In [ ]:
python -m smtag.datagen.ann2th -L1200 -X5 -f 5X_L1200_disease compendium/ncbi_disease/train

# B. Moving training data to Amazon EFS

In [9]:
cd ../cloud

In [ ]:
scp -i basicLinuxAMI.pem 5X_L1200_small_molecule_train.zip ec2-user@smtag-web:/efs/smtag/data/

In [ ]:
scp -i basicLinuxAMI.pem 5X_L1200_gene_protein_train.zip ec2-user@smtag-web:/efs/smtag/data/

In [ ]:
scp -i basicLinuxAMI.pem 5X_L1200_subcell_train.zip ec2-user@smtag-web:/efs/smtag/data/

In [ ]:
scp -i basicLinuxAMI.pem 5X_L1200_cell_train.zip ec2-user@smtag-web:/efs/smtag/data/

In [ ]:
scp -i basicLinuxAMI.pem 5X_L1200_tissue_train.zip ec2-user@smtag-web:/efs/smtag/data/

In [ ]:
scp -i basicLinuxAMI.pem 5X_L1200_organism_train.zip ec2-user@smtag-web:/efs/smtag/data/

In [ ]:
scp -i basicLinuxAMI.pem 5X_L1200_exp_assay_train.zip ec2-user@smtag-web:/efs/smtag/data/

In [ ]:
scp -i basicLinuxAMI.pem 5X_L1200_disease_train.zip ec2-user@smtag-web:/efs/smtag/data/

# C. Training models on multi-GPU

## Entity models

### Small molecule model
120 epoch with learning rate 0.01 and batch size 128

In [ ]:
python meta -E120 -Z128 -R0.01 -o small_molecule -f 5X_L1200_small_molecule_train

### Gene product model

### Cellular component model

### Cell model

### Tissue and organ model

### Organism model

### Experimental assay model

### Disease model 

## Roles

### Roles of gene products

### Roles of small molecules

### Reporter gene products

In [1]:
python -m smtag.train.meta -f 5X_L1200_small_molecule_w_train -E120 -R0.01 -Z128 

opt[namebase]=5X_L1200_small_molecule_w_train
opt[learning_rate]=0.01
opt[epochs]=120
opt[minibatch_size]=128
opt[selected_features]=['geneprod']
opt[collapsed_features]=[]
opt[overlap_features]=[]
opt[features_as_input]=[]
opt[nf_table]=[8, 8, 8]
opt[pool_table]=[2, 2, 2]
opt[kernel_table]=[6, 6, 6]
opt[dropout]=0.1
opt[validation_fraction]=0.2
nf.output= 1
nf.input= 32
index_of_collapsed_feature= None
index_of_overlap_feature None
concept2index self.selected_features [21]
concept2index self.collapsed_features []
concept2index self.overlap_features []
Loading ./data/5X_L1200_small_molecule_w_train.pyth as features for the dataset.
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/lemberger/Documents/code/py-smtag/s

: 1

2. Benchmark on the test set, without tokenization (option `-T`)

In [ ]:
python -m smtag.train.evaluate -f 5X_L1200_small_molecule_w_test -m 5X_L1200_small_molecule -T

# D. Copy models locally

# E. Evaluate models with test data